File: LeetCode/5회차/B15-206-Reverse Linked List.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        res, node = None, head

        while node:
            res, res.next, node = node, res, node.next

        return res


x = [1,2,3,4,5]
root = cur = ListNode()
for num in x:
    cur.next = ListNode(num)
    cur = cur.next
print(Solution().reverseList(root.next))

File: LeetCode/JavaScript/B15-206-Reverse Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const reverse = (node, prev = null) => {
  if (!node) {
    return prev;
  }
  const next = node.next;
  node.next = prev;
  return reverse(next, node);
};

var reverseList = function (head) {
  return reverse(head);
};
